src/lib/drivers/accelerometer/PX4Accelerometer.cpp

```diff
@@ -0,0 +1,150 @@
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+
+#include "PX4Accelerometer.hpp"
+
+#include <lib/drivers/device/Device.hpp>
+
+PX4Accelerometer::PX4Accelerometer(uint32_t device_id, uint8_t priority, enum Rotation rotation) :
+	CDev(nullptr),
+	ModuleParams(nullptr),
+	_sensor_accel_pub{ORB_ID(sensor_accel), priority},
+	_rotation{get_rot_matrix(rotation)}
+{
+	_class_device_instance = register_class_devname(ACCEL_BASE_DEVICE_PATH);
+
+	_sensor_accel_pub.get().device_id = device_id;
+	_sensor_accel_pub.get().scaling = 1.0f;
+
+	// set software low pass filter for controllers
+	updateParams();
+	configure_filter(_filter_cutoff.get());
+
+	// force initial publish to allocate uORB buffer
+	// TODO: can be removed once all drivers are in threads
+	_sensor_accel_pub.update();
+}
+
+PX4Accelerometer::~PX4Accelerometer()
+{
+	if (_class_device_instance != -1) {
+		unregister_class_devname(ACCEL_BASE_DEVICE_PATH, _class_device_instance);
+	}
+}
+
+int PX4Accelerometer::ioctl(cdev::file_t *filp, int cmd, unsigned long arg)
+{
+	switch (cmd) {
+	case ACCELIOCSSCALE: {
+			// Copy offsets and scale factors in
+			accel_calibration_s cal{};
+			memcpy(&cal, (accel_calibration_s *) arg, sizeof(cal));
+
+			_calibration_offset = matrix::Vector3f{cal.x_offset, cal.y_offset, cal.z_offset};
+			_calibration_scale = matrix::Vector3f{cal.x_scale, cal.y_scale, cal.z_scale};
+		}
+
+		return PX4_OK;
+
+	case DEVIOCGDEVICEID:
+		return _sensor_accel_pub.get().device_id;
+
+	default:
+		return -ENOTTY;
+	}
+}
+
+void PX4Accelerometer::set_device_type(uint8_t devtype)
+{
+	// current DeviceStructure
+	union device::Device::DeviceId device_id;
+	device_id.devid = _sensor_accel_pub.get().device_id;
+
+	// update to new device type
+	device_id.devid_s.devtype = devtype;
+
+	// copy back to report
+	_sensor_accel_pub.get().device_id = device_id.devid;
+}
+
+void PX4Accelerometer::update(hrt_abstime timestamp, int16_t x, int16_t y, int16_t z)
+{
+	sensor_accel_s &report = _sensor_accel_pub.get();
```

Comment:
This feels wrong. I see what the motivation is, but the getter is used to be the setter. 
.ref() my feel better.

Is the timestamp stored in the report but not published suggesting something or does all this belong in the if? If it is the latter comment it.

Reply:
I agree it's a bit weird, but already a fairly common pattern. The Publication class wraps the a copy of the structure and the orb handling (advertise and publish). 

I would propose we review the uORB::Publication wrappers directly and update the interface to be clear and hard to misuse. https://github.com/PX4/Firmware/blob/master/src/modules/uORB/Publication.hpp